src/lib/mina_base/parties_logic.ml

```diff
@@ -0,0 +1,441 @@
+                   ; protocol_state_predicate=
+                       start_data.protocol_state_predicate
+                   ; other_parties= remaining })
+            in
+            Transaction_commitment.if_ is_start' ~then_:on_start
+              ~else_:local_state.transaction_commitment
+      in
+      let local_state =
+        { local_state with
+          transaction_commitment
+        ; token_id=
+            Token_id.if_ is_start' ~then_:Token_id.default
+              ~else_:local_state.token_id }
+      in
+      ((party, remaining), local_state)
+    in
+    let local_state =
+      { local_state with
+        parties= remaining
+      ; will_succeed=
+          ( match is_start with
+          | `Yes start_data ->
+              start_data.will_succeed
+          | `No ->
+              local_state.will_succeed
+          | `Compute start_data ->
+              Bool.if_ is_start' ~then_:start_data.will_succeed
+                ~else_:local_state.will_succeed ) }
+    in
+    let a, inclusion_proof =
+      h.perform (Get_account (party, local_state.ledger))
+    in
+    h.perform (Check_inclusion (local_state.ledger, a, inclusion_proof)) ;
+    let predicate_satisfied : Bool.t =
+      h.perform (Check_predicate (is_start', party, a, global_state))
+    in
+    let a', update_permitted =
+      h.perform
+        (Check_auth_and_update_account
+           { is_start= is_start'
+           ; global_state
+           ; party
+           ; account= a
+           ; transaction_commitment= local_state.transaction_commitment
+           ; inclusion_proof })
+    in
+    let party_succeeded =
+      Bool.(
+        protocol_state_predicate_satisfied && predicate_satisfied
+        && update_permitted)
+    in
+    (* The first party must succeed. *)
+    Bool.(assert_ ((not is_start') || party_succeeded)) ;
+    let local_state =
+      { local_state with
+        success= Bool.( && ) local_state.success party_succeeded }
+    in
+    let local_delta =
+      Amount.(h.perform (Balance a) - h.perform (Balance a'))
+    in
+    let party_token = h.perform (Party_token_id party) in
+    let fee_excess_change0, new_local_fee_excess =
+      let curr_token : Token_id.t = local_state.token_id in
+      let same_token = Token_id.equal curr_token party_token in
+      let curr_is_default = Token_id.(equal default) curr_token in
+      let should_merge = Bool.( && ) (Bool.not same_token) curr_is_default in
+      let to_merge_amount =
+        Amount.if_ should_merge ~then_:local_state.excess ~else_:Amount.zero
+      in
+      let new_local =
+        let base =
+          Amount.if_ same_token ~then_:local_state.excess ~else_:Amount.zero
+        in
+        Amount.add_signed base local_delta
+      in
+      (to_merge_amount, new_local)
+    in
+    let local_state = {local_state with excess= new_local_fee_excess} in
+    let global_state =
+      (* TODO: Maybe overflows should be possible and cause a transaction failure? *)
```

Comment:
This scenario can be triggered with non-default tokens, so we'll have to keep them disabled until this is fixed or handled elsewhere.

Reply:
We only need to keep them disabled for fee payment right?

Reply:
And actually I don't think there's any risk here as I check that the token is the default one before adding it to the excess.